### PR TITLE
Add `Fp64`

### PR DESCRIPTION
**Pull request by Pratyush**
## Description

Adds `Fp64`.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: ff/src/fields/arithmetic.rs [hunk 1]
```diff
@@ -80,22 +94,28 @@ macro_rules! impl_field_into_repr {
...
-            let mut all_bits_set = P::MODULUS.0[$limbs - 1] == !0 - (1 << 63);
-            for i in 1..$limbs {
-                all_bits_set &= P::MODULUS.0[$limbs - i - 1] == core::u64::MAX;
-            }
-            let _no_carry: bool = !(first_bit_set || all_bits_set);
-
             #[cfg(use_asm)]
             #[allow(unsafe_code, unused_mut)]
             {
+                // Checking the modulus at compile time
```
Comment (Pratyush) [inline ff/src/fields/arithmetic.rs:111]: Drive-by refactor that moves code used only when `use_asm` is true inside the conditionally-compiled portion.

File: ff/src/fields/arithmetic.rs [hunk 2]
```diff
@@ -5,12 +5,26 @@
 /// `P::MODULUS` has (a) a non-zero MSB, and (b) at least one
 /// zero bit in the rest of the modulus.
 macro_rules! impl_field_mul_assign {
+    (1) => {
+        fn mul_assign(&mut self, other: &Self) {
+            #[inline(always)]
+            fn mont_reduce(&mut self, mul_result: u128) {
```
Comment (Pratyush) [inline ff/src/fields/arithmetic.rs:11]: This is just a specialization of the standard algorithm to the case where there is just one limb.

File: ff/src/fields/arithmetic.rs [hunk 3]
```diff
@@ -120,8 +121,11 @@ macro_rules! impl_field_square_in_place {
             }
             r[$limbs * 2 - 1] = r[$limbs * 2 - 2] >> 63;
             for i in 0..$limbs {
-                r[$limbs * 2 - 2 - i] =
-                    (r[$limbs * 2 - 2 - i] << 1) | (r[$limbs * 2 - 3 - i] >> 63);
+                let subtractor = (2 * ($limbs - 1usize))
+                    .checked_sub(i + 1)
+                    .map(|index| r[index])
+                    .unwrap_or(0);
```
Comment (Pratyush) [inline ff/src/fields/arithmetic.rs:127]: Required to avoid underflowing. I have confirmed locally that it doesn't affect performance.

File: ff/src/fields/arithmetic.rs [hunk 4]
```diff
@@ -201,14 +205,20 @@ macro_rules! impl_prime_field_standard_sample {
...
-                default_int.0[1] = upper;
+                if $limbs == 1 {
+                    default_int.0[0] = (other % u128::from(P::MODULUS.0[0])) as u64;
+                } else {
+                    let upper = (other >> 64) as u64;
+                    let lower = ((other << 64) >> 64) as u64;
+                    let limbs = [lower, upper];
+                    for (cur, other) in default_int.0.iter_mut().zip(&limbs) {
+                        *cur = *other;
+                    }
```
Comment (Pratyush) [inline ff/src/fields/arithmetic.rs:220]: Again, required to satisfy the compiler.

File: ff/src/fields/arithmetic.rs [hunk 5]
```diff
@@ -11,6 +11,7 @@ macro_rules! impl_field_mul_assign {
         fn mul_assign(&mut self, other: &Self) {
             // Checking the modulus at compile time
             let first_bit_set = P::MODULUS.0[$limbs - 1] >> 63 != 0;
+            #[allow(unused_mut)]
```
Comment (ValarDragon) [inline ff/src/fields/arithmetic.rs:14]:
```suggestion
            // $limbs can be 1, hence we can run into a case with an unused mut.
            #[allow(unused_mut)]
```

File: ff/src/fields/arithmetic.rs [hunk 6]
```diff
@@ -22,7 +23,7 @@ macro_rules! impl_field_mul_assign {
                 #[cfg(use_asm)]
                 #[allow(unsafe_code, unused_mut)]
                 {
-                    if $limbs <= 6 {
+                    if $limbs <= 6 && $limbs > 1 {
```
Comment (ValarDragon) [inline ff/src/fields/arithmetic.rs:26]: Not related to this PR, but could there be a comment for why theres a different method for limbs > 6? Is it just that the larger limb size case isn't implemented? (Doesn't need to block this PR)

File: ff/src/fields/arithmetic.rs (hunk 3 above)
Comment (ValarDragon) [reply ff/src/fields/arithmetic.rs:127]: Can a comment be added for the arithmetic operations these lines are doing / intending to do? (Looks like propagating a carry bit?)

I agree this matches what was previously implemented.

File: ff/src/fields/arithmetic.rs [hunk 7]
```diff
@@ -201,14 +205,20 @@ macro_rules! impl_prime_field_standard_sample {
...
-                let lower = ((other << 64) >> 64) as u64;
                 let mut default_int = P::BigInt::default();
-                default_int.0[0] = lower;
-                default_int.0[1] = upper;
+                if $limbs == 1 {
+                    default_int.0[0] = (other % u128::from(P::MODULUS.0[0])) as u64;
+                } else {
+                    let upper = (other >> 64) as u64;
+                    let lower = ((other << 64) >> 64) as u64;
+                    let limbs = [lower, upper];
```
Comment (ValarDragon) [inline ff/src/fields/arithmetic.rs:217]:
```suggestion
                    // This is equivalent to the following, but satisfying the compiler:
                    // default_int.0[0] = lower;
                    // default_int.0[1] = upper;
                    let limbs = [lower, upper];
```

**Review by ValarDragon — COMMENTED** [review]
LGTM, left a couple of requests for additional comments, then I think its good to merge

File: ff/src/fields/arithmetic.rs (hunk 3 above)
Comment (Pratyush) [reply ff/src/fields/arithmetic.rs:137]: I'm not too sure myself, but I added a comment explaining what is computed.

File: ff/src/fields/arithmetic.rs (hunk 6 above)
Comment (Pratyush) [reply ff/src/fields/arithmetic.rs:28]: Yes, larger limb size is not implemented

File: ff/src/fields/arithmetic.rs (hunk 3 above)
Comment (ValarDragon) [reply ff/src/fields/arithmetic.rs:137]: Ok cool!

Kinda worrisome if neither of us understand the code, we should file an issue to decipher & document whats going on these functions 😅

**Review by ValarDragon — APPROVED** [review]
LGTM

File: ff/src/fields/arithmetic.rs (hunk 3 above)
Comment (Pratyush) [reply ff/src/fields/arithmetic.rs:137]: Agreed, @jon-chuang should have the best state, I believe?

**Comment by Pratyush** [comment]
This passes tests locally for the prime 257.
